File: src/agentic_template_kit/lockfile.py
from __future__ import annotations

import hashlib
from pathlib import Path

import yaml

from .models import RenderedFile

LOCKFILE = ".agentic-template.lock"


def sha256_text(value: str) -> str:
    return "sha256:" + hashlib.sha256(value.encode("utf-8")).hexdigest()


def write_lockfile(target: Path, files: list[RenderedFile], target_name: str) -> None:
    data = {
        "version": "1",
        "target": target_name,
        "managed_files": [
            {
                "path": str(rendered.destination),
                "platform": rendered.platform,
                "source": rendered.source,
                "checksum": sha256_text(rendered.content),
            }
            for rendered in files
        ],
    }
    (target / LOCKFILE).write_text(yaml.safe_dump(data, sort_keys=False), encoding="utf-8")


def load_lockfile(target: Path) -> dict:
    path = target / LOCKFILE
    if not path.exists():
        return {}
    return yaml.safe_load(path.read_text(encoding="utf-8")) or {}

File: internal/lockfile/lockfile.go
package lockfile

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"

	"github.com/agentic-template-kit/skcr/internal/models"
	"gopkg.in/yaml.v3"
)

const LockfileName = ".agentic-template.lock"

func Sha256Text(value string) string {
	hash := sha256.Sum256([]byte(value))
	return "sha256:" + hex.EncodeToString(hash[:])
}

func WriteLockfile(targetDir string, files []models.RenderedFile, targetName string) error {
	managed := make([]map[string]any, 0, len(files))
	for _, rendered := range files {
		managed = append(managed, map[string]any{
			"path":     rendered.Destination,
			"platform": rendered.Platform,
			"source":   rendered.Source,
			"checksum": Sha256Text(rendered.Content),
		})
	}
	payload := map[string]any{
		"version":       "1",
		"target":        targetName,
		"managed_files": managed,
	}
	encoded, err := yaml.Marshal(payload)
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(targetDir, LockfileName), encoded, 0o644)
}

func LoadLockfile(targetDir string) (map[string]any, error) {
	path := filepath.Join(targetDir, LockfileName)
	if _, err := os.Stat(path); err != nil {
		if os.IsNotExist(err) {
			return map[string]any{}, nil
		}
		return nil, err
	}
	payload, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	decoded := map[string]any{}
	if err := yaml.Unmarshal(payload, &decoded); err != nil {
		return nil, err
	}
	if decoded == nil {
		return map[string]any{}, nil
	}
	return decoded, nil
}

func ManagedFilesByPath(lock map[string]any) map[string]map[string]any {
	result := map[string]map[string]any{}
	raw, ok := lock["managed_files"].([]any)
	if !ok {
		return result
	}
	for _, item := range raw {
		entry, ok := item.(map[string]any)
		if !ok {
			continue
		}
		path, _ := entry["path"].(string)
		if path == "" {
			continue
		}
		result[path] = entry
	}
	return result
}
